node/src/proxy_server/mod.rs

```diff
@@ -923,31 +939,43 @@ impl ProxyServer {
         }
     }
 
-    fn get_return_route_info(
+    fn rri_from_remaining_route(
         &self,
         remaining_route: &Route,
-        source: &str,
-    ) -> Option<Rc<AddReturnRouteMessage>> {
+    ) -> Option<u32> {
         let mut mut_remaining_route = remaining_route.clone();
         mut_remaining_route
             .shift(self.main_cryptde)
             .expect("Internal error: remaining route in ProxyServer with no hops");
-        let return_route_id = match mut_remaining_route.id(self.main_cryptde) {
-            Ok(rri) => rri,
+        match mut_remaining_route.id(self.main_cryptde) {
+            Ok(rri) => Some(rri),
             Err(e) => {
                 error!(self.logger, "Can't report services consumed: {}", e);
-                return None;
-            }
-        };
-        match self.route_ids_to_return_routes.get(&return_route_id) {
-            Some(rri) => Some(rri),
-            None => {
-                error!(self.logger, "Can't report services consumed: received response with bogus return-route ID {} for {}. Ignoring", return_route_id, source);
                 None
             }
         }
     }
 
+    fn get_return_route_info(
+        &mut self,
+        return_route_id: u32,
+        source: &str,
+    ) -> Option<Rc<AddReturnRouteMessage>> {
+        match self.route_ids_to_return_routes_first_chance.remove(&return_route_id) {
+            Some(rri) => {
+                self.route_ids_to_return_routes_stragglers.insert(return_route_id, (*rri).clone());
+                Some(rri)
+            },
+            None => match self.route_ids_to_return_routes_stragglers.get(&return_route_id) {
```

Comment:
If we don't find matching RRI in the `_first_chance` map, we immediately look in the `_stragglers` map.